Fix the implementation of JSTraceable for RefCell.

eefriedman opened the pull request:
The existing implementation could panic; make sure that doesn't
happen by requiring that the contents of a RefCell are trivially
traceable (i.e. the value don't contain any traceable objects).

I'm not sure whether the TriviallyJSTraceable trait is actually
worthwhile; maybe we should just never use RefCell in the DOM.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8056)

<!-- Reviewable:end -->

frewsxcv commented:
Regression test?

eefriedman commented:
This was found by inspection; I'm not sure it's actually possible to trigger this panic with Servo on master.  The relevant sequence would be something like "call `RefCell::borrow_mut()`, then trigger a GC with the borrow active".

DOMRefCell sort of has the same issue, but the documentation for that explicitly acknowledges the problem, and it has a workaround.

jdm commented:
We've attempted to keep RefCell out of script/dom/ in the past, but never got around to actually enforcing it via a lint.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8041) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8091) made this pull request unmergeable. Please resolve the merge conflicts.

eefriedman commented:
Rebased.  (Should this be assigned to someone?)

jdm commented:
I would rather remove all uses of RefCell (and the implementation of JSTraceable for RefCell) than add another layer of annotations for "this type doesn't contain any GC-owned values". That's the purpose of `no_jsmanaged_fields`, which I believe is clearer.

eefriedman committed:
Remove JSTraceable implementation from RefCell.

The implementation wasn't really right, and we would rather just use
DOMRefCell anyway.

eefriedman commented:
Okay; here's the version which just makes RefCell untraceable.

jdm commented:
@bors-servo: r+
Thanks!
<!-- Reviewable comment -K29RZ2oqa0qvvPUW5BE:-1746383326 -->

---

Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8056)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit df7fb8f has been approved by `jdm`

<!-- @bors-servo r=jdm df7fb8fa326e2b061e2da8c833cc558273db5f37 -->

bors-servo commented:
:hourglass: Testing commit df7fb8f with merge 4f51710...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/177)

eefriedman commented:
@bors-servo retry #7787

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/165), [gonk](http://build.servo.org/builders/gonk/builds/170), [linux-dev](http://build.servo.org/builders/linux-dev/builds/163), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/176), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/175), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/185) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/177)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/165), [gonk](http://build.servo.org/builders/gonk/builds/170), [linux-dev](http://build.servo.org/builders/linux-dev/builds/163), [linux-rel](http://build.servo.org/builders/linux-rel/builds/178), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/176), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/175), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/185)
